packages/cardano-services/src/Program/utils.ts

```diff
@@ -0,0 +1,250 @@
+import { RabbitMqTxSubmitProvider } from '@cardano-sdk/rabbitmq';
+import { ServiceNames } from './ServiceNames';
+import { WebSocketClosed, ogmiosTxSubmitProvider, urlToConnectionConfig } from '@cardano-sdk/ogmios';
+import Logger from 'bunyan';
+import dns, { SrvRecord } from 'dns';
+import pRetry, { FailedAttemptError } from 'p-retry';
+
+export const SERVICE_DISCOVERY_BACKOFF_FACTOR_DEFAULT = 1.1;
+export const SERVICE_DISCOVERY_TIMEOUT_DEFAULT = 60 * 1000;
+export const DNS_SRV_CACHE_KEY = 'DNS_SRV_Record';
+
+export type RetryBackoffConfig = {
+  factor?: number;
+  maxRetryTime?: number;
+};
+
+export const onFailedAttemptFor =
+  (serviceName: string, logger: Logger) =>
+  async ({ attemptNumber, message, retriesLeft }: FailedAttemptError) => {
+    const nextAction = retriesLeft > 0 ? 'retrying...' : 'exiting';
+    logger.trace(message);
+    logger.debug(
+      `Establishing connection to ${serviceName}: Attempt ${attemptNumber} of ${
+        attemptNumber + retriesLeft
+      }, ${nextAction}`
+    );
+    if (retriesLeft === 0) {
+      logger.error(message);
+      // Invokes onDeath() callback within cardano-services entrypoints, following by server.shutdown() and process.exit(1)
+      process.kill(process.pid, 'SIGTERM');
+    }
+  };
+
+// Select the first random record from the DNS server resolved list
+export const resolveSrvRecord = async (serviceName: string): Promise<SrvRecord> => {
+  const [srvRecord] = await dns.promises.resolveSrv(serviceName);
+  return srvRecord;
+};
+
+export const createDnsResolver = (config: RetryBackoffConfig, logger: Logger) => async (serviceName: string) =>
+  await pRetry(async () => await resolveSrvRecord(serviceName), {
+    factor: config.factor,
+    maxRetryTime: config.maxRetryTime,
+    onFailedAttempt: onFailedAttemptFor(serviceName, logger)
+  });
+
+export type DnsResolver = ReturnType<typeof createDnsResolver>;
+
+/**
+ * Creates a extended Pool client :
+ * - use passed srv service name in order to resolve the port
+ * - make dealing with failovers (re-resolving the port) opaque
+ * - use exponential backoff retry internally with default timeout and factor
+ * - intercept 'query' operation and handle connection errors runtime
+ * - all other operations are bind to pool object withoud modifications
+ *
+ * @returns pg.Pool instance
+ */
+export const getPoolWithServiceDiscovery = async (
+  dnsResolver: DnsResolver,
+  { host, database, password, user }: ClientConfig
+): Promise<Pool> => {
+  const { name, port } = await dnsResolver(host!);
+  let pool = new Pool({ database, host: name, password, port, user });
+
+  return new Proxy<Pool>({} as Pool, {
+    get(_, prop) {
+      if (prop === 'then') return;
+      if (prop === 'query') {
+        return (args: string | QueryConfig, values?: any) =>
+          pool.query(args, values).catch(async (error) => {
+            if (error.code && ['ENOTFOUND', 'ECONNREFUSED'].includes(error.code)) {
```

Comment:
I'm wondering if we also need to watch for `ECONNRESET` :thinking:

Reply:
I've seen only the `ENOTFOUND`/ `ECONNREFUSED` during local testing, but sounds reasonable to watch for `ECONNRESET` as well just in case (`A connection was forcibly closed by a peer`). Will add it

Reply:
8fd7c5267f8250848f954350f58bfef3735aea12